src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -57,6 +57,12 @@ public override void Dispose()
                     _sslStream.Dispose();
                     _sslStream = null;
                 }
+
+                if (_tcpStream != null)
+                {
+                    _tcpStream.Dispose();
```

Comment:
_stream only ever stores _tcpStream or _sslStream, which are both disposed here (if not disposed already). 

Increased async safety in EnableSsl/DisableSsl can be handled in a separate improvement.